File: src/vectorpin/verifier.py
# Copyright 2025 Jascha Wanger / Tarnover, LLC
# SPDX-License-Identifier: Apache-2.0
"""Pin verification.

Three failure modes, in order of severity:

  1. SignatureInvalid — the producer is not who the pin claims, or the
     attestation has been re-signed by an attacker.
  2. VectorTampered — the vector in the store does not match what the
     producer attested to. This is the steganography kill shot.
  3. SourceMismatch — the source text the verifier is checking against
     does not match what the producer pinned. Either the verifier has
     the wrong text, or the source corpus drifted.

The Verifier returns a structured VerificationResult so callers can
distinguish these and route them differently (alert vs. quarantine vs.
re-pin).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

import numpy as np
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from vectorpin.attestation import PROTOCOL_VERSION, Pin
from vectorpin.hash import hash_text, hash_vector


class VerifyError(Enum):
    """Distinct failure modes that callers can route on."""

    OK = "ok"
    UNKNOWN_KEY = "unknown_key"
    UNSUPPORTED_VERSION = "unsupported_version"
    SIGNATURE_INVALID = "signature_invalid"
    VECTOR_TAMPERED = "vector_tampered"
    SOURCE_MISMATCH = "source_mismatch"
    MODEL_MISMATCH = "model_mismatch"
    SHAPE_MISMATCH = "shape_mismatch"


@dataclass(frozen=True)
class VerificationResult:
    """Structured result. Truthy iff verification succeeded."""

    ok: bool
    error: VerifyError
    detail: str = ""

    def __bool__(self) -> bool:
        return self.ok


def _public_key_from_bytes(raw: bytes) -> Ed25519PublicKey:
    return Ed25519PublicKey.from_public_bytes(raw)


class Verifier:
    """Verifies Pin attestations against a key registry.

    The registry maps key id -> public key. Verifiers MUST be willing to
    hold multiple keys at once to support rotation: when a new signing
    key is introduced, both the old and new public keys live in the
    registry until the rotation window closes.
    """

    def __init__(self, public_keys: dict[str, Ed25519PublicKey | bytes]):
        self._keys: dict[str, Ed25519PublicKey] = {}
        for kid, key in public_keys.items():
            if isinstance(key, bytes):
                self._keys[kid] = _public_key_from_bytes(key)
            elif isinstance(key, Ed25519PublicKey):
                self._keys[kid] = key
            else:
                raise TypeError(f"public key for {kid!r} must be Ed25519PublicKey or bytes")

    def add_key(self, kid: str, key: Ed25519PublicKey | bytes) -> None:
        """Register an additional public key — used during rotation."""
        if isinstance(key, bytes):
            key = _public_key_from_bytes(key)
        self._keys[kid] = key

    def verify(
        self,
        pin: Pin,
        *,
        source: str | None = None,
        vector: np.ndarray | None = None,
        expected_model: str | None = None,
    ) -> VerificationResult:
        """Verify a Pin against optional ground-truth source/vector.

        The signature check always runs. The other checks run only when
        the corresponding ground truth is supplied — letting callers do
        partial verification (e.g. signature-only when the source text
        is unavailable but the producer identity still matters).
        """
        if pin.header.v != PROTOCOL_VERSION:
            return VerificationResult(
                False,
                VerifyError.UNSUPPORTED_VERSION,
                f"pin version {pin.header.v} not supported by this verifier",
            )

        public_key = self._keys.get(pin.kid)
        if public_key is None:
            return VerificationResult(
                False,
                VerifyError.UNKNOWN_KEY,
                f"no registered public key for kid={pin.kid!r}",
            )

        try:
            public_key.verify(pin.sig, pin.header.canonicalize())
        except InvalidSignature:
            return VerificationResult(
                False,
                VerifyError.SIGNATURE_INVALID,
                "ed25519 signature did not verify",
            )

        if vector is not None:
            if vector.ndim != 1 or vector.shape[0] != pin.header.vec_dim:
                return VerificationResult(
                    False,
                    VerifyError.SHAPE_MISMATCH,
                    f"vector shape {vector.shape} does not match pin (dim={pin.header.vec_dim})",
                )
            if hash_vector(vector, pin.header.vec_dtype) != pin.header.vec_hash:
                return VerificationResult(
                    False,
                    VerifyError.VECTOR_TAMPERED,
                    "vector hash mismatch — embedding has been modified after pinning",
                )

        if source is not None and hash_text(source) != pin.header.source_hash:
            return VerificationResult(
                False,
                VerifyError.SOURCE_MISMATCH,
                "source hash mismatch — pinned source differs from supplied source",
            )

        if expected_model is not None and pin.header.model != expected_model:
            return VerificationResult(
                False,
                VerifyError.MODEL_MISMATCH,
                f"pin model {pin.header.model!r} != expected {expected_model!r}",
            )

        return VerificationResult(True, VerifyError.OK)

File: rust/vectorpin/src/verifier.rs
// Copyright 2025 Jascha Wanger / Tarnover, LLC
// SPDX-License-Identifier: Apache-2.0

//! Pin verification.
//!
//! Mirrors the Python verifier: same failure-mode enum, same matching
//! semantics, same support for partial verification (signature-only,
//! signature + vector, etc).

use std::collections::HashMap;

use ed25519_dalek::{Signature, Verifier as _, VerifyingKey};

use crate::attestation::{Pin, PROTOCOL_VERSION};
use crate::hash::{hash_text, hash_vector, VecDtype, VectorRef};

/// Distinct verification failure modes. Callers route on this so a
/// signature-invalid result (potential forgery) can be handled
/// differently from a vector-tampered result (potential steganography
/// kill shot).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VerifyError {
    /// Pin uses a protocol version this verifier does not understand.
    UnsupportedVersion(u32),
    /// `kid` not present in the verifier's key registry.
    UnknownKey(String),
    /// Ed25519 signature did not verify against the canonical header.
    SignatureInvalid,
    /// Vector hash mismatch — the embedding has been modified after pinning.
    VectorTampered,
    /// Source text hash mismatch.
    SourceMismatch,
    /// Pin was issued for a different model than the caller expected.
    ModelMismatch {
        /// Model identifier in the pin.
        pin_model: String,
        /// Model identifier the caller asked us to require.
        expected: String,
    },
    /// Supplied vector's dim did not match the dim in the pin header.
    ShapeMismatch {
        /// Length of the supplied vector.
        supplied: usize,
        /// `vec_dim` from the pin header.
        expected: u32,
    },
    /// Pin failed to parse one of its dtype-related fields.
    UnsupportedDtype(String),
}

impl std::fmt::Display for VerifyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            VerifyError::UnsupportedVersion(v) => write!(f, "unsupported pin version: {v}"),
            VerifyError::UnknownKey(k) => write!(f, "unknown signing key id: {k}"),
            VerifyError::SignatureInvalid => write!(f, "ed25519 signature did not verify"),
            VerifyError::VectorTampered => write!(
                f,
                "vector hash mismatch — embedding has been modified after pinning"
            ),
            VerifyError::SourceMismatch => write!(f, "source hash mismatch"),
            VerifyError::ModelMismatch {
                pin_model,
                expected,
            } => {
                write!(f, "pin model {pin_model:?} != expected {expected:?}")
            }
            VerifyError::ShapeMismatch { supplied, expected } => {
                write!(
                    f,
                    "vector shape mismatch: supplied len {supplied}, pin dim {expected}"
                )
            }
            VerifyError::UnsupportedDtype(s) => write!(f, "unsupported canonical dtype: {s}"),
        }
    }
}

impl std::error::Error for VerifyError {}

/// Holds the public-key registry for one or more `kid` values and runs
/// pin verification against supplied ground truth.
#[derive(Default)]
pub struct Verifier {
    keys: HashMap<String, VerifyingKey>,
}

impl Verifier {
    /// Construct an empty verifier. Add public keys with [`Self::add_key`].
    pub fn new() -> Self {
        Self::default()
    }

    /// Register a public key under `kid`. Multiple keys may live in
    /// the registry simultaneously to support rotation.
    pub fn add_key(&mut self, kid: &str, public_key_bytes: [u8; 32]) {
        if let Ok(vk) = VerifyingKey::from_bytes(&public_key_bytes) {
            self.keys.insert(kid.to_owned(), vk);
        }
    }

    /// Number of registered keys (sanity check for tests).
    pub fn key_count(&self) -> usize {
        self.keys.len()
    }

    /// Verify just the signature — useful when ground-truth source/vector
    /// are unavailable but producer identity still matters.
    pub fn verify_signature(&self, pin: &Pin) -> Result<(), VerifyError> {
        if pin.header.v != PROTOCOL_VERSION {
            return Err(VerifyError::UnsupportedVersion(pin.header.v));
        }
        let key = self
            .keys
            .get(&pin.kid)
            .ok_or_else(|| VerifyError::UnknownKey(pin.kid.clone()))?;
        let sig_bytes: [u8; 64] = pin
            .sig
            .as_slice()
            .try_into()
            .map_err(|_| VerifyError::SignatureInvalid)?;
        let signature = Signature::from_bytes(&sig_bytes);
        key.verify(&pin.header.canonicalize(), &signature)
            .map_err(|_| VerifyError::SignatureInvalid)
    }

    /// Verify signature + any supplied ground truth.
    ///
    /// Pass `Some(...)` for whichever components you have on hand:
    /// the signature is always checked; source/vector/model checks run
    /// only when the corresponding argument is present. This mirrors
    /// the Python reference verifier so callers can do partial
    /// verification (e.g. signature-only at retrieval time, full
    /// verification at audit time).
    pub fn verify_full<'a, V>(
        &self,
        pin: &Pin,
        source: Option<&str>,
        vector: Option<V>,
        expected_model: Option<&str>,
    ) -> Result<(), VerifyError>
    where
        V: Into<VectorRef<'a>>,
    {
        self.verify_signature(pin)?;

        if let Some(vec) = vector {
            let vec = vec.into();
            if vec.len() as u32 != pin.header.vec_dim {
                return Err(VerifyError::ShapeMismatch {
                    supplied: vec.len(),
                    expected: pin.header.vec_dim,
                });
            }
            let dtype = VecDtype::parse(&pin.header.vec_dtype)
                .map_err(|_| VerifyError::UnsupportedDtype(pin.header.vec_dtype.clone()))?;
            if hash_vector(vec, dtype) != pin.header.vec_hash {
                return Err(VerifyError::VectorTampered);
            }
        }

        if let Some(s) = source {
            if hash_text(s) != pin.header.source_hash {
                return Err(VerifyError::SourceMismatch);
            }
        }

        if let Some(em) = expected_model {
            if pin.header.model != em {
                return Err(VerifyError::ModelMismatch {
                    pin_model: pin.header.model.clone(),
                    expected: em.to_owned(),
                });
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::signer::Signer;

    fn fixture(kid: &str) -> (Signer, Verifier, Vec<f32>) {
        let signer = Signer::generate(kid.into());
        let mut verifier = Verifier::new();
        verifier.add_key(signer.key_id(), signer.public_key_bytes());
        let v: Vec<f32> = (0..16).map(|i| (i as f32) * 0.1).collect();
        (signer, verifier, v)
    }

    #[test]
    fn verify_full_passes_on_honest_inputs() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("hello", "m", v.as_slice()).unwrap();
        verifier
            .verify_full(&pin, Some("hello"), Some(v.as_slice()), None)
            .expect("honest verify must succeed");
    }

    #[test]
    fn verify_signature_only_passes() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("hello", "m", v.as_slice()).unwrap();
        verifier.verify_signature(&pin).unwrap();
    }

    #[test]
    fn vector_tamper_is_caught() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("hello", "m", v.as_slice()).unwrap();
        let mut tampered = v.clone();
        tampered[0] += 1e-5;
        let err = verifier
            .verify_full(&pin, None::<&str>, Some(tampered.as_slice()), None)
            .unwrap_err();
        assert_eq!(err, VerifyError::VectorTampered);
    }

    #[test]
    fn source_mismatch_is_caught() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("hello", "m", v.as_slice()).unwrap();
        let err = verifier
            .verify_full(&pin, Some("HELLO"), None::<&[f32]>, None)
            .unwrap_err();
        assert_eq!(err, VerifyError::SourceMismatch);
    }

    #[test]
    fn unknown_key_is_caught() {
        let signer = Signer::generate("rogue".into());
        let v: Vec<f32> = vec![1.0, 2.0, 3.0];
        let pin = signer.pin("x", "m", v.as_slice()).unwrap();
        let other = Signer::generate("prod".into());
        let mut verifier = Verifier::new();
        verifier.add_key(other.key_id(), other.public_key_bytes());
        let err = verifier.verify_signature(&pin).unwrap_err();
        assert!(matches!(err, VerifyError::UnknownKey(_)));
    }

    #[test]
    fn shape_mismatch_is_caught() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("x", "m", v.as_slice()).unwrap();
        let truncated: Vec<f32> = v.iter().take(8).copied().collect();
        let err = verifier
            .verify_full(&pin, None::<&str>, Some(truncated.as_slice()), None)
            .unwrap_err();
        assert!(matches!(err, VerifyError::ShapeMismatch { .. }));
    }

    #[test]
    fn model_mismatch_is_caught() {
        let (signer, verifier, v) = fixture("k1");
        let pin = signer.pin("x", "model-A", v.as_slice()).unwrap();
        let err = verifier
            .verify_full(&pin, None::<&str>, None::<&[f32]>, Some("model-B"))
            .unwrap_err();
        assert!(matches!(err, VerifyError::ModelMismatch { .. }));
    }

    #[test]
    fn key_rotation_works() {
        let old = Signer::generate("2026-04".into());
        let new = Signer::generate("2026-05".into());
        let mut verifier = Verifier::new();
        verifier.add_key(old.key_id(), old.public_key_bytes());
        verifier.add_key(new.key_id(), new.public_key_bytes());
        let v: Vec<f32> = vec![1.0, 2.0];
        verifier
            .verify_signature(&old.pin("x", "m", v.as_slice()).unwrap())
            .unwrap();
        verifier
            .verify_signature(&new.pin("x", "m", v.as_slice()).unwrap())
            .unwrap();
    }
}
